tool/embedded-binaries.ts

```diff
@@ -70,162 +71,216 @@ export async function getEmbeddedProtocol(outPath: string): Promise<void> {
+      ${p.join(outPath, repo)}`,
+      {silent: true}
+    );
+  }
+
+  const version = ref ? `commit ${ref}` : 'latest update';
+  console.log(`Fetching ${version} for ${repo}.`);
+  shell.exec(`git fetch --depth=1 origin ${ref ?? 'master'}`, {
+    silent: true,
+    cwd: p.join(outPath, repo),
+  });
+  shell.exec('git reset --hard FETCH_HEAD', {
+    silent: true,
+    cwd: p.join(outPath, repo),
+  });
+}
+
+// Links the built files at `builtPath` into `outPath`.
+async function linkBuiltFiles(
+  repo: string,
+  builtPath: string,
+  outPath: string
+) {
+  console.log(`Linking built ${repo} into ${outPath}.`);
+  await cleanEmbeddedDir(outPath, repo);
+  if (getOs() === 'windows') {
+    shell.cp('-R', builtPath, p.join(outPath, repo));
+  } else {
+    // Symlinking doesn't play nice with Jasmine's test globbing on Windows.
+    fs.symlink(p.resolve(builtPath), p.join(outPath, repo));
   }
 }
 
-// Ensures that `outPath` exists and its `sass_embedded` subdirectory does not.
-async function cleanEmbeddedDir(outPath: string): Promise<void> {
+// Builds the embedded proto at `protoPath` into a pbjs with TS declaration
+// file.
+function buildEmbeddedProtocol(protoPath: string) {
+  console.log(`Building pbjs and TS declaration file from ${protoPath}.`);
+  try {
+    const protocPath =
+      getOs() === 'windows'
+        ? '%CD%/node_modules/protoc/protoc/bin/protoc.exe'
+        : 'node_modules/protoc/protoc/bin/protoc';
+    const pluginPath =
+      getOs() === 'windows'
+        ? '%CD%/node_modules/.bin/protoc-gen-ts.cmd'
+        : 'node_modules/.bin/protoc-gen-ts';
+    shell.exec(
+      `${protocPath} \
+      --plugin="protoc-gen-ts=${pluginPath}" \
+      --js_out="import_style=commonjs,binary:." \
+      --ts_out="." \
+      ${protoPath}`,
+      {silent: true}
+    );
+  } catch (error) {
+    throw Error(`Failed to write proto to pbjs: ${error.message}.`);
+  }
+}
+
+// Builds the Embedded Dart Sass executable from the source at `repoPath`.
+function buildDartSassEmbedded(repoPath: string) {
+  console.log('Downloading dart-sass-embedded dependencies.');
+  shell.exec('pub upgrade', {
+    cwd: repoPath,
+    silent: true,
+  });
+
+  console.log('Building dart-sass-embedded executable.');
+  shell.exec('pub run grinder protobuf pkg-standalone-dev', {
+    cwd: repoPath,
+    silent: true,
+  });
+}
+
+// Ensures that `outPath` exists and the `subdir` does not.
+async function cleanEmbeddedDir(
+  outPath: string,
+  subdir: string
+): Promise<void> {
   await fs.mkdir(outPath, {recursive: true});
   try {
-    await fs.rmdir(p.join(outPath, 'sass_embedded'), {recursive: true});
+    await fs.rmdir(p.join(outPath, subdir), {recursive: true});
   } catch (_) {
     // If the path doesn't exist yet, that's fine.
   }
 }
 
 // Gets the current platform's operating system. Throws if the operating system
 // is not supported by Dart Sass Embedded.
-function getOs(): string {
+function getOs(): 'linux' | 'macos' | 'windows' {
```

Comment:
Note: Since the OS is going to be constant for the entire run of the script, you could potentially make this (and arch/file extension) constants rather than functions.